okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/FastFallbackExchangeFinder.kt

```diff
@@ -35,132 +35,139 @@ internal class FastFallbackExchangeFinder(
 ) {
   private val connectDelayMillis = 250L
 
-  /** Plans currently being connected, and that will later be added to [connectResults]. */
-  private var connectsInFlight = CopyOnWriteArrayList<Plan>()
+  /**
+   * Plans currently being connected, and that will later be added to [connectResults]. This is
+   * mutated by the call thread only. If is accessed by background connect threads.
+   */
+  private val tcpConnectsInFlight = CopyOnWriteArrayList<Plan>()
+
+  /** Plans we started, and canceled, because another plan won a race. */
+  private val deferredPlans = ArrayDeque<Plan>()
 
   /**
    * Results are posted here as they occur. The find job is done when either one plan completes
    * successfully or all plans fail.
    */
   private val connectResults = taskRunner.backend.decorate(LinkedBlockingDeque<ConnectResult>())
 
-  /** Exceptions accumulate here. */
-  private var firstException: IOException? = null
-
   /** True until we've launched all the connects we'll ever launch. */
-  private var morePlansExist = true
+  private var routePlannerHasMoreRoutes = true
 
   fun find(): RealConnection {
+    var firstException: IOException? = null
     try {
-      while (morePlansExist || connectsInFlight.isNotEmpty()) {
+      while (routePlannerHasMoreRoutes ||
+        tcpConnectsInFlight.isNotEmpty() ||
+        deferredPlans.isNotEmpty()
+      ) {
         if (routePlanner.isCanceled()) throw IOException("Canceled")
 
-        launchConnect()
+        launchTcpConnect()
 
-        val connection = awaitConnection()
-        if (connection != null) return connection
+        var connectResult = awaitTcpConnect() ?: continue
 
-        morePlansExist = morePlansExist && routePlanner.hasMoreRoutes()
-      }
+        if (connectResult.isSuccess) {
+          // We have a connected TCP connection. Cancel the other racing connects that all lost.
+          cancelInFlightConnects()
 
-      throw firstException!!
-    } finally {
-      for (plan in connectsInFlight) {
-        plan.cancel()
+          // Finish connecting. We won't have to if the winner is from the connection pool.
+          if (!connectResult.plan.isReady) {
```

Comment:
Is it better if this doesn't need to be checked? Can connectTlsEtc just shortcircuit and return this?

Reply:
Yeah I like that, but when I tried to change to it it gets a bit sloppy. For example, in the FakePlan it’ll need to identify whether to sleep a 2nd time.